rustvmm_gen/lib_kernel_source.py

```diff
@@ -0,0 +1,136 @@
+    except requests.exceptions.RequestException as e:
+        return (False, f"Network error: {str(e)}")
+
+
+def create_temp_dir(version):
+    temp_dir = os.path.join("/tmp", f"linux-{version}-prepare")
+
+    if os.path.exists(temp_dir):
+        shutil.rmtree(temp_dir)
+
+    try:
+        os.makedirs(temp_dir)
+        return temp_dir
+    except OSError as e:
+        raise RuntimeError(f"Failed to create temp directory: {e}") from e
+
+
+def download_kernel(version, temp_dir):
+    version_major = re.match(r"^(\d+)\.\d+(\.\d+)?$", version).group(1)
+    url = f"{KERNEL_ORG_CDN}/v{version_major}.x/linux-{version}.tar.xz"
+    tarball_path = os.path.join(temp_dir, f"linux-{version}.tar.xz")
+    print(f"Downloading {url} to {tarball_path}")
+
+    try:
+        with requests.get(url, stream=True) as response:
+            response.raise_for_status()
+            total_size = int(response.headers.get("content-length", 0))
+            downloaded = 0
+
+            with open(tarball_path, "wb") as f:
+                for chunk in response.iter_content(chunk_size=8192):
+                    f.write(chunk)
+                    downloaded += len(chunk)
+                    if total_size > 0:
+                        progress = downloaded / total_size * 100
+                        print(f"\rDownloading: {progress:.1f}%", end="")
+                print()
+        return tarball_path
+    except Exception as e:
+        raise RuntimeError(f"Download failed: {e}") from e
+
+
+def extract_kernel(tarball_path, temp_dir):
+    print("Extracting...")
+    try:
+        with tarfile.open(tarball_path, "r:xz") as tar:
+            tar.extractall(path=temp_dir)
+            extract_path = os.path.join(
+                temp_dir, f"{os.path.basename(tarball_path).split('.tar')[0]}"
+            )
+            print(f"Extracted to {extract_path}")
+            return extract_path
+    except (tarfile.TarError, IOError) as e:
+        raise RuntimeError(f"Extraction failed: {e}") from e
+
+
+def install_headers(src_dir, arch, install_path, cleanup=False):
+    os.chdir(src_dir)
+
+    parent_dir = os.path.dirname(src_dir)
+    if install_path is None:
+        install_path = os.path.join(parent_dir, f"{arch}_headers")
+
+    try:
+        os.makedirs(install_path, exist_ok=True)
+
+        abs_install_path = os.path.abspath(install_path)
+        print(f"Installing to {abs_install_path}")
+        result = subprocess.run(
+            [
+                "make",
+                "headers_install",
+                f"ARCH={arch}",
+                f"INSTALL_HDR_PATH={abs_install_path}",
+            ],
+            check=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            text=True,
+        )
+        print(result.stdout)
+
+        if cleanup:
+            print(f"Cleaning temporary files at {parent_dir}")
+            shutil.rmtree(parent_dir)
+
+        return True
```

Comment:
I'd also just drop the boolean return value here and simply raise an exception in the failure case